clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -314,69 +370,116 @@ public void poll(long timeout) throws IOException {
+                keysHandled.add(key);
+                KafkaChannel channel = channel(key);
+
+                // register all per-connection metrics at once
+                sensors.maybeRegisterConnectionMetrics(channel.id());
+                if (idleExpiryManager != null)
+                    idleExpiryManager.update(channel.id(), currentTimeNanos);
+
+                try {
+
+                    /* complete any connections that have finished their handshake (either normally or immediately) */
+                    if (isImmediatelyConnected || key.isConnectable()) {
+                        madeProgress = true;
+                        if (channel.finishConnect()) {
+                            this.connected.add(channel.id());
+                            this.sensors.connectionCreated.record();
+                            SocketChannel socketChannel = (SocketChannel) key.channel();
+                            log.debug("Created socket with SO_RCVBUF = {}, SO_SNDBUF = {}, SO_TIMEOUT = {} to node {}",
+                                    socketChannel.socket().getReceiveBufferSize(),
+                                    socketChannel.socket().getSendBufferSize(),
+                                    socketChannel.socket().getSoTimeout(),
+                                    channel.id());
+                        } else
+                            continue;
+                    }
 
-                /* complete any connections that have finished their handshake (either normally or immediately) */
-                if (isImmediatelyConnected || key.isConnectable()) {
-                    if (channel.finishConnect()) {
-                        this.connected.add(channel.id());
-                        this.sensors.connectionCreated.record();
-                        SocketChannel socketChannel = (SocketChannel) key.channel();
-                        log.debug("Created socket with SO_RCVBUF = {}, SO_SNDBUF = {}, SO_TIMEOUT = {} to node {}",
-                                socketChannel.socket().getReceiveBufferSize(),
-                                socketChannel.socket().getSendBufferSize(),
-                                socketChannel.socket().getSoTimeout(),
-                                channel.id());
-                    } else
-                        continue;
-                }
+                    /* if channel is not ready finish prepare */
+                    if (channel.isConnected() && !channel.ready()) {
+                        madeProgress = true;
+                        channel.prepare();
+                    }
 
-                /* if channel is not ready finish prepare */
-                if (channel.isConnected() && !channel.ready())
-                    channel.prepare();
+                    /* if channel is ready read from any connections that have readable data */
+                    if (channel.ready() && key.isReadable() && !hasStagedReceive(channel)) {
```

Comment:
will do. since explicitlyMutedChannels lookups, while being O(1) are hash-based, and mutating explicitlyMutedChannels is O(log(#requests-in-progress)) which happens twice per request, do you think its better to replace it with an extra boolean flag on channel? have boolean muted and boolean explicitelyMuted? (or rather bool mutedForOOM and bool mutedForOrdering)